crypto/bio/bio_tfo.h

```diff
@@ -0,0 +1,122 @@
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * Contains definitions for simplifying the use of TCP Fast Open
+ * in OpenSSL socket BIOs.
+ */
+
+/* If a supported OS is added here, update test/bio_tfo_test.c */
+#if defined(TCP_FASTOPEN) && !defined(OPENSSL_NO_TFO)
+
+# if defined(OPENSSL_SYS_MACOSX) || defined(__FreeBSD__)
+#  include <sys/sysctl.h>
+# endif
+
+/*
+ * OSSL_TFO_SYSCTL is used to determine if TFO is supported by
+ * this kernel, and if supported, if it is enabled. This is more of
+ * a problem on FreeBSD 10.3 ~ 12.0, where TCP_FASTOPEN was defined,
+ * but not enabled by default in the kernel.
+ * Linux does not have sysctlbyname(), and the closest equivalent
+ * is to go into the /proc filesystem, but I'm not sure it's
+ * worthwhile.
+ *
+ * The OSSL_TFO_xxxxxx_FLAGs can be OR'd together:
+ * 0 = TFO disabled
+ * 1 = client TFO enabled
+ * 2 = server TFO enabled
+ * 3 = server and client TFO enabled
+ */
+
+# define OSSL_TFO_CLIENT_FLAG 1
+# define OSSL_TFO_SERVER_FLAG 2
+
+/*
+ * Some options are purposely NOT defined per-platform
+ *
+ * OSSL_TFO_SYSCTL
+ *     Defined as a sysctlbyname() option to to determine if
+ *     TFO is enabled in the kernel (macOS, FreeBSD)
+ *
+ * OSSL_TFO_SERVER_SOCKOPT
+ *     Defined to indicate the socket option used to enable
+ *     TFO on a server socket (all)
+ *
+ * OSSL_TFO_SERVER_SOCKOPT_VALUE
+ *     Value to be used with OSSL_TFO_SERVER_SOCKOPT
+ *
+ * OSSL_TFO_CONNECTX
+ *     Use the connectx() function to make a client connection
+ *     (macOS)
+ *
+ * OSSL_TFO_CLIENT_SOCKOPT
+ *     Defined to indicate the socket option used to enable
+ *     TFO on a client socket (FreeBSD, Linux 4.14 and later)
+ *
+ * OSSL_TFO_SENDTO
+ *     Defined to indicate the sendto() message type to
+ *     be used to initiate a TFO connection (FreeBSD,
+ *     Linux pre-4.14)
+ *
+ * OSSL_TFO_DO_NOT_CONNECT
+ *     Defined to skip calling conect() when creating a
+ *     client socket (macOS, FreeBSD, Linux pre-4.14)
+ */
+
+# if defined(OPENSSL_SYS_WINDOWS)
+/*
+ * NO WINDOWS SUPPORT
+ *
+ * But this is is what would be used on the server:
+ *
+ * define OSSL_TFO_SERVER_SOCKOPT       TCP_FASTOPEN
+ * define OSSL_TFO_SERVER_SOCKOPT_VALUE 1
+ *
+ * Still have to figure out client support
+ */
+#  undef TCP_FASTOPEN
+# endif
+
+/* NO VMS SUPPORT */
+# if defined(OPENSSL_SYS_VMS)
+#  undef TCP_FASTOPEN
+# endif
+
+# if defined(OPENSSL_SYS_MACOSX)
+#  define OSSL_TFO_SYSCTL               "net.inet.tcp.fastopen"
+#  define OSSL_TFO_SERVER_SOCKOPT       TCP_FASTOPEN
+#  define OSSL_TFO_SERVER_SOCKOPT_VALUE 1
+#  define OSSL_TFO_CONNECTX             1
+#  define OSSL_TFO_DO_NOT_CONNECT       1
+# endif
+
+# if defined(__FreeBSD__)
+#  define OSSL_TFO_SYSCTL               "net.inet.tcp.fastopen.enabled"
```

Comment:
Yeah, I'm re-working BSD a little bit, since the support changed at 12.0.0; this will take more time (as I have to now install two versions of FreeBSD somewhere...)